R/cube-residuals.R

```diff
@@ -23,46 +67,185 @@ standardizedMRResiduals <- function(cube) {
 #'
-#' Standardized residuals, \code{(observed - expected) / sqrt(V)}, where \code{V}
-#' is the residual cell variance (Agresti, 2007, section 2.4.5). Special care is taken
-#' for multiple-response variables which are in effect a series of separate tables where
-#' \sQuote{not selected} cells for each item are are hidden.
+#' Comparing a column or row with a baseline column or row. This calculates the
+#' z-score for the cells when comparing `x` to the baseline columns
 #'
-#' @param model A CrunchCube representing a contingency table
+#' @param cube a cube to calculate the comparison on
+#' @param dim which dimension is being compared (`rows` or `cols`, only valid
+#'   for `compareDims()`)
+#' @param baseline a character, the column to use as a baseline to compare `x`
+#'   against
+#' @param x a character, the column to compare against the baseline
+#' @param ... arguments passed from `compareRows()` or `compareCols()` to
+#'   `compareDims()` (i.e. `baseline` and `x`)
 #'
-#' @return an array of standardized residuals or Z-scores from the hypothesis being tested.
-#' The default method is that the joint distributions of (weighted) counts are equal to the
-#' marginal distributions of the table.
+#' @return the z-score for the column or row given in `x`
 #'
-#' @seealso [stats::chisq.test]
-#' @references
-#' Agresti, A. (2007)
-#' \emph{An Introduction to Categorical Data Analysis, 2nd ed.},
-#' New York: John Wiley & Sons. Page 38.
-#' @importFrom stats chisq.test
-#' @name cube-residuals
-#' @aliases rstandard,CrunchCube-method cube-residuals
+#' @name dimension-comparison
 NULL
 
-#' @rdname cube-residuals
+#' @rdname dimension-comparison
 #' @export
-setMethod("rstandard", "CrunchCube", function(model) {
-    if (length(dimensions(model)) > 2) {
-        halt(
-            "Cannot compute residuals with more than two dimensions. Pick a ",
-            "slice to evaluate."
-        )
+compareCols <- function(cube, ...) compareDims(cube = cube, dim = "cols", ...)
+
+#' @rdname dimension-comparison
+#' @export
+compareRows <- function(cube, ...) compareDims(cube = cube, dim = "rows", ...)
+
+#' @rdname dimension-comparison
+#' @export
+compareDims <- function(cube, dim = c("cols", "rows"), baseline, x) {
+    dim <- match.arg(dim)
+
+    # grab the names of extents for each dimensions
+    # TODO: we shouldn't need to do this if we can just pass baseline/x into the
+    # subsetting function when we can subset by name _or_ index natively
+    if (dim == "cols") {
+        names <- colnames(as.array(cube))
+    } else if (dim == "rows") {
+        names <- rownames(as.array(cube))
     }
 
-    if (any(startsWith(getDimTypes(model), "mr_"))) {
-        return(standardizedMRResiduals(model))
+    # convert to numeric indices
+    # TODO: remove when we can accept either characters or indices and pass them
+    # to subsetting unmolested
+    if (is.character(baseline)) {
+        baseline_ind <- which(names == baseline)
```

Comment:
Will this cause problems for duplicated or out of order character vectors? It might be easier to error until we merge in the cube-subset enhancement PR

Reply:
These shouldn't be vectors, so I've added in `length == 1` here (and added tests), in which case duplicated / out of order doesn't matter.